## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
arr_t1 = np.random.randint(-8, 8, size=[10])

print("Оригинальный массив:")
print(arr_t1)

arr_t1[(arr_t1 > 3) & (arr_t1 < 8)] *= -1

print("\nМассив после изменения знаков:")
print(arr_t1)

Оригинальный массив:
[ 4  5  4  5 -5 -8  3  1 -2  5]

Массив после изменения знаков:
[-4 -5 -4 -5 -5 -8  3  1 -2 -5]


**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
arr_t2 = np.random.randint(-20, 20, size=[20])

print("Оригинальный массив:")
print(arr_t2)

print(f"\nМаксимальный элемент: {np.max(arr_t2)}")

arr_t2[np.argmax(arr_t2)] = 0

print("\nМассив после изменения:")
print(arr_t2)

Оригинальный массив:
[ -1  13  12  16   6  10 -11   5 -16  11  17   9  -6   7 -16  18  19   5
  -3  11]

Максимальный элемент: 19

Массив после изменения:
[ -1  13  12  16   6  10 -11   5 -16  11  17   9  -6   7 -16  18   0   5
  -3  11]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [4]:
arr_t3 = np.arange(0, 12).reshape(6, 2)

print("Оригинальный массив:")
print(arr_t3)

grids_t3 = np.meshgrid(*arr_t3)

cartesian_product_t3 = np.array(grids_t3).T.reshape(-1, len(arr_t3))

print("\nПрямое произведение массивов:")
print(cartesian_product_t3)

Оригинальный массив:
[[ 0  1]
 [ 2  3]
 [ 4  5]
 [ 6  7]
 [ 8  9]
 [10 11]]

Прямое произведение массивов:
[[ 0  2  4  6  8 10]
 [ 0  3  4  6  8 10]
 [ 1  2  4  6  8 10]
 [ 1  3  4  6  8 10]
 [ 0  2  5  6  8 10]
 [ 0  3  5  6  8 10]
 [ 1  2  5  6  8 10]
 [ 1  3  5  6  8 10]
 [ 0  2  4  7  8 10]
 [ 0  3  4  7  8 10]
 [ 1  2  4  7  8 10]
 [ 1  3  4  7  8 10]
 [ 0  2  5  7  8 10]
 [ 0  3  5  7  8 10]
 [ 1  2  5  7  8 10]
 [ 1  3  5  7  8 10]
 [ 0  2  4  6  9 10]
 [ 0  3  4  6  9 10]
 [ 1  2  4  6  9 10]
 [ 1  3  4  6  9 10]
 [ 0  2  5  6  9 10]
 [ 0  3  5  6  9 10]
 [ 1  2  5  6  9 10]
 [ 1  3  5  6  9 10]
 [ 0  2  4  7  9 10]
 [ 0  3  4  7  9 10]
 [ 1  2  4  7  9 10]
 [ 1  3  4  7  9 10]
 [ 0  2  5  7  9 10]
 [ 0  3  5  7  9 10]
 [ 1  2  5  7  9 10]
 [ 1  3  5  7  9 10]
 [ 0  2  4  6  8 11]
 [ 0  3  4  6  8 11]
 [ 1  2  4  6  8 11]
 [ 1  3  4  6  8 11]
 [ 0  2  5  6  8 11]
 [ 0  3  5  6  8 11]
 [ 1  2  5  6  8 11]
 [ 1  3  5  6  8 11]
 [ 0  2  4  7  8 11]
 [ 0  3  4  7  8 11]
 [ 1  2  4 

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [5]:
arr_a_t4 = np.array([[1, 4, 3], [9, 5, 6], [12, 8, 9], [2, 4, 7], 
                    [12, 9, 8], [0, 6, 9], [1, 12, 9], [8, 5, 9]])
arr_b_t4 = np.array([[8,5], [9,12]])

print("Массив А:")
print(arr_a_t4)
print("\nМассив B:")
print(arr_b_t4)

def contains_elements(a_row: np.array, b_row_1: np.array, b_row_2: np.array) -> bool:
    return np.any(np.isin(a_row, b_row_1)) and np.any(np.isin(a_row, b_row_2))

mask_t4 = np.array([contains_elements(row, arr_b_t4[0], arr_b_t4[1]) for row in arr_a_t4])

print("\nСтроки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B:")
print(arr_a_t4[mask_t4])

Массив А:
[[ 1  4  3]
 [ 9  5  6]
 [12  8  9]
 [ 2  4  7]
 [12  9  8]
 [ 0  6  9]
 [ 1 12  9]
 [ 8  5  9]]

Массив B:
[[ 8  5]
 [ 9 12]]

Строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B:
[[ 9  5  6]
 [12  8  9]
 [12  9  8]
 [ 8  5  9]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [6]:
arr_t5 = np.array([
    [0, 0, 0],
    [3, 7, 9],
    [3, 5, 2],
    [4, 4, 4],
    [4, 2, 6],
    [6, 7, 8],
    [1, 6, 7],
    [5, 3, 1],
    [1, 1, 1],
    [6, 2, 6]
])

print("Оригинальный массив:")
print(arr_t5)

def are_row_elems_equal(row: np.array) -> bool:
    return np.all(row == row[0])

mask_t5 = np.array([not are_row_elems_equal(a_row) for a_row in arr_t5])

print("Массив со строками из неравных значений:")
print(arr_t5[mask_t5])

Оригинальный массив:
[[0 0 0]
 [3 7 9]
 [3 5 2]
 [4 4 4]
 [4 2 6]
 [6 7 8]
 [1 6 7]
 [5 3 1]
 [1 1 1]
 [6 2 6]]
Массив со строками из неравных значений:
[[3 7 9]
 [3 5 2]
 [4 2 6]
 [6 7 8]
 [1 6 7]
 [5 3 1]
 [6 2 6]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [7]:
arr_t6 = np.array([
    [4, 4, 4],
    [3, 7, 9],
    [3, 5, 2],
    [4, 4, 4],
    [4, 4, 4],
    [6, 2, 6],
    [1, 6, 7],
    [5, 3, 1],
    [1, 1, 1],
    [6, 2, 6]
])

print("Оригинальный массив:")
print(arr_t6)

print("Массив после удаления повторяющихся строк:")
print(np.unique(arr_t6, axis=0))

Оригинальный массив:
[[4 4 4]
 [3 7 9]
 [3 5 2]
 [4 4 4]
 [4 4 4]
 [6 2 6]
 [1 6 7]
 [5 3 1]
 [1 1 1]
 [6 2 6]]
Массив после удаления повторяющихся строк:
[[1 1 1]
 [1 6 7]
 [3 5 2]
 [3 7 9]
 [4 4 4]
 [5 3 1]
 [6 2 6]]


______
______

**7.** Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [8]:
# Pure Python (PP)
arr_t71_pp = [
    [0, 0, 1, 9, 8],
    [2, 4, 2, 6, 1],
    [3, 0, 0, 1, 2],
    [4, 4, 4, 25, 6]
]
print("Pure Python")
print("Матрица:")
for row in arr_t71_pp:
    print(row)
    
rows_t71_pp = len(arr_t71_pp)
cols_t71_pp = len(arr_t71_pp[0]) if rows_t71_pp > 0 else 0

dim_size_t71_pp = min(rows_t71_pp, cols_t71_pp)

product_t71_pp = 1

for i in range(dim_size_t71_pp):
    if arr_t71_pp[i][i] != 0:
        product_t71_pp *= arr_t71_pp[i][i]

print(f"Произведение ненулевых элементов на диагонали прямоугольной матрицы: {product_t71_pp}")

Pure Python
Матрица:
[0, 0, 1, 9, 8]
[2, 4, 2, 6, 1]
[3, 0, 0, 1, 2]
[4, 4, 4, 25, 6]
Произведение ненулевых элементов на диагонали прямоугольной матрицы: 100


In [9]:
# NumPy (NP)
arr_t71_np = np.array([
    [0, 0, 1, 9, 8],
    [2, 4, 2, 6, 1],
    [3, 0, 0, 1, 2],
    [4, 4, 4, 25, 6]
])

print("NumPy")
print("Матрица:")
print(arr_t71_np)

diag_t71_np = np.diagonal(arr_t71_np)

nonzero_diag_t71_np = diag_t71_np[diag_t71_np != 0]
product_t71_np = np.prod(nonzero_diag_t71_np)

print(f"Произведение ненулевых элементов на диагонали прямоугольной матрицы: {product_t71_np}")

NumPy
Матрица:
[[ 0  0  1  9  8]
 [ 2  4  2  6  1]
 [ 3  0  0  1  2]
 [ 4  4  4 25  6]]
Произведение ненулевых элементов на диагонали прямоугольной матрицы: 100


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [10]:
# Pure Python (PP)
from collections import Counter

x_t72_pp = [1,3,3,3,5,5]
y_t72_pp = [3,1,5,3,3,5]
z_t72_pp = [3,3,3,3,3,3]

print("Pure Python")
print(f"Вектор Х: {x_t72_pp}")
print(f"Вектор Y: {y_t72_pp}")
print(f"Вектор Z: {z_t72_pp}")

print(f"Задают ли вектора X и Y мультимножество: {Counter(x_t72_pp) == Counter(y_t72_pp)}")
print(f"Задают ли вектора X и Z мультимножество: {Counter(x_t72_pp) == Counter(z_t72_pp)}")
print(f"Задают ли вектора Y и Z мультимножество: {Counter(y_t72_pp) == Counter(z_t72_pp)}")

Pure Python
Вектор Х: [1, 3, 3, 3, 5, 5]
Вектор Y: [3, 1, 5, 3, 3, 5]
Вектор Z: [3, 3, 3, 3, 3, 3]
Задают ли вектора X и Y мультимножество: True
Задают ли вектора X и Z мультимножество: False
Задают ли вектора Y и Z мультимножество: False


In [11]:
#NumPy (NP)
x_t72_np = np.array([1,3,3,3,5,5])
y_t72_np = np.array([3,1,5,3,3,5])
z_t72_np = np.array([3,3,3,3,3,3])

print("NumPy")
print(f"Вектор Х: {x_t72_np}")
print(f"Вектор Y: {y_t72_np}")
print(f"Вектор Z: {z_t72_np}")

x_t72_np_unique, x_t72_np_counts = np.unique(x_t72_np, return_counts=True)
y_t72_np_unique, y_t72_np_counts = np.unique(y_t72_np, return_counts=True)
z_t72_np_unique, z_t72_np_counts = np.unique(z_t72_np, return_counts=True)


print(f"Задают ли вектора X и Y мультимножество: {np.array_equal(x_t72_np_unique, y_t72_np_unique) and np.array_equal(x_t72_np_counts, y_t72_np_counts)}")
print(f"Задают ли вектора X и Z мультимножество: {np.array_equal(x_t72_np_unique, z_t72_np_unique) and np.array_equal(x_t72_np_counts, z_t72_np_counts)}")
print(f"Задают ли вектора Y и Z мультимножество: {np.array_equal(y_t72_np_unique, z_t72_np_unique) and np.array_equal(y_t72_np_counts, z_t72_np_counts)}")

NumPy
Вектор Х: [1 3 3 3 5 5]
Вектор Y: [3 1 5 3 3 5]
Вектор Z: [3 3 3 3 3 3]
Задают ли вектора X и Y мультимножество: True
Задают ли вектора X и Z мультимножество: False
Задают ли вектора Y и Z мультимножество: False


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [12]:
# Pure Python (PP)
arr_t73_pp = [6, 2, 0, 3, 0, 0, 5, 7, 0, 11, 157, 0, 9, 0, 0, 155, 0, 1, 789, 156]

print("Pure Python")
print("Вектор:")
print(arr_t73_pp)

max_val_t73_pp = None

for i in range(1, len(arr_t73_pp)):
    if arr_t73_pp[i-1] == 0:
        if max_val_t73_pp is None or arr_t73_pp[i] > max_val_t73_pp:
            max_val_t73_pp = arr_t73_pp[i]

print(f"Максимальный элемент в векторе, перед которыми стоит ноль: {max_val_t73_pp}")

Pure Python
Вектор:
[6, 2, 0, 3, 0, 0, 5, 7, 0, 11, 157, 0, 9, 0, 0, 155, 0, 1, 789, 156]
Максимальный элемент в векторе, перед которыми стоит ноль: 155


In [13]:
# NumPy(NP)
arr_t73_np = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0, 11, 157, 0, 9, 0, 0, 155, 0, 1, 789, 156])

print("NumPy")
print("Вектор:")
print(arr_t73_np)

max_val_t73_np = None

shifted_zero_mask_t73_np = np.roll(arr_t73_np == 0, shift=1)
elements_after_zero_t73_np = arr_t73_np[shifted_zero_mask_t73_np]

if elements_after_zero_t73_np.size != 0:
    max_val_t73_np = np.max(elements_after_zero_t73_np)
    
print(f"Максимальный элемент в векторе, перед которыми стоит ноль: {max_val_t73_np}")

NumPy
Вектор:
[  6   2   0   3   0   0   5   7   0  11 157   0   9   0   0 155   0   1
 789 156]
Максимальный элемент в векторе, перед которыми стоит ноль: 155


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [14]:
# Pure Python (PP)
arr_t74_pp = [2, 2, 2, 3, 3, 3, 5, 2, 2, 10, 5, 5, 5, 3, 3, 3, 10]

print("Pure Python")
print("Вектор:")
print(arr_t74_pp)

values_t74_pp, counts_t74_pp = [],[]

current_value_t74_pp = arr_t74_pp[0]
count_t74_pp = 1

for i in range(1, len(arr_t74_pp)):
    if arr_t74_pp[i] == current_value_t74_pp:
        count_t74_pp += 1
    else:
        values_t74_pp.append(current_value_t74_pp)
        counts_t74_pp.append(count_t74_pp)
        current_value_t74_pp = arr_t74_pp[i]
        count_t74_pp = 1
        
values_t74_pp.append(current_value_t74_pp)
counts_t74_pp.append(count_t74_pp)

print("Кодировка длин серий в векторе:")
print(f"Значения: {values_t74_pp}")
print(f"Повторы: {counts_t74_pp}")

Pure Python
Вектор:
[2, 2, 2, 3, 3, 3, 5, 2, 2, 10, 5, 5, 5, 3, 3, 3, 10]
Кодировка длин серий в векторе:
Значения: [2, 3, 5, 2, 10, 5, 3, 10]
Повторы: [3, 3, 1, 2, 1, 3, 3, 1]


In [15]:
# NumPy (NP)
arr_t74_np = np.array([2, 2, 2, 3, 3, 3, 5, 2, 2, 10, 5, 5, 5, 3, 3, 3, 10])

print("NumPy")
print("Вектор:")
print(arr_t74_np)

diff_indices_t74_np = np.where(np.diff(arr_t74_np) != 0)[0] + 1

values_t74_np = arr_t74_np[np.hstack(([0], diff_indices_t74_np))]

counts_t74_np = np.diff(np.hstack(([0], diff_indices_t74_np, [len(arr_t74_np)])))

print("Кодировка длин серий в векторе:")
print(f"Значения: {values_t74_np}")
print(f"Повторы: {counts_t74_np}")

NumPy
Вектор:
[ 2  2  2  3  3  3  5  2  2 10  5  5  5  3  3  3 10]
Кодировка длин серий в векторе:
Значения: [ 2  3  5  2 10  5  3 10]
Повторы: [3 3 1 2 1 3 3 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [51]:
# Pure Python
import timeit
x_t75_pp = [[0, -1, 3], [9, -8, 0], [1, -1, 34], [45, -2, 6], [9, -4, -7]]
y_t75_pp = [[16, 9, 5], [26, -26, 13], [4, 4, 4], [0, 0, 37], [77, 5, 2]]

print("Pure Python")
print(f"Объеты выборки X: {x_t75_pp}")
print(f"Объекты выборки Y: {y_t75_pp}")

def euclidean_distance_pp(x_objects, y_objects):
    distances = []
    for x in x_objects:
        row = []
        for y in y_objects:
            dist = sum((xi-yi) ** 2 for xi, yi in zip(x,y)) ** 0.5
            row.append(dist)
        distances.append(row)
    return distances

print(f"Матрица евклидовых расстояний: {euclidean_distance_pp(x_t75_pp, y_t75_pp)}")
print(f"Время выполнения 100000 повторений: {timeit.timeit(lambda: euclidean_distance_pp(x_t75_pp, y_t75_pp), number= 100000)}")

Pure Python
Объеты выборки X: [[0, -1, 3], [9, -8, 0], [1, -1, 34], [45, -2, 6], [9, -4, -7]]
Объекты выборки Y: [[16, 9, 5], [26, -26, 13], [4, 4, 4], [0, 0, 37], [77, 5, 2]]
Матрица евклидовых расстояний: [[18.973665961010276, 37.429934544425805, 6.48074069840786, 34.0147027033899, 77.23988606931007], [19.05255888325765, 27.964262908219126, 13.601470508735444, 38.91015291668744, 69.26037828369117], [34.14674215792775, 41.12177038990418, 30.56141357987225, 3.3166247903554, 82.68010643437755], [31.0322412983658, 31.400636936215164, 41.48493702538308, 54.680892457969264, 33.0], [19.026297590440446, 34.249087579087416, 14.491376746189438, 45.088801270381985, 69.18092222571191]]
Время выполнения 100000 повторений: 2.4661054000025615


In [52]:
#NumPy
x_t75_np = np.array([[0, -1, 3], [9, -8, 0], [1, -1, 34], [45, -2, 6], [9, -4, -7]])
y_t75_np = np.array([[16, 9, 5], [26, -26, 13], [4, 4, 4], [0, 0, 37], [77, 5, 2]])

print("NumPy")
print(f"Объеты выборки X:")
print(x_t75_np)
print(f"Объекты выборки Y:")
print(y_t75_np)

def euclidean_distance_np(x_objects, y_objects):
    diff = x_objects[:, np.newaxis, :] - y_objects
    return np.sqrt(np.sum(diff**2, axis=2))

print(f"Матрица евклидовых расстояний:\n {euclidean_distance_np(x_t75_np, y_t75_np)}")
print(f"Время выполнения 100000 повторений: {timeit.timeit(lambda: euclidean_distance_np(x_t75_np, y_t75_np), number= 100000)}")

NumPy
Объеты выборки X:
[[ 0 -1  3]
 [ 9 -8  0]
 [ 1 -1 34]
 [45 -2  6]
 [ 9 -4 -7]]
Объекты выборки Y:
[[ 16   9   5]
 [ 26 -26  13]
 [  4   4   4]
 [  0   0  37]
 [ 77   5   2]]
Матрица евклидовых расстояний:
 [[18.97366596 37.42993454  6.4807407  34.0147027  77.23988607]
 [19.05255888 27.96426291 13.60147051 38.91015292 69.26037828]
 [34.14674216 41.12177039 30.56141358  3.31662479 82.68010643]
 [31.0322413  31.40063694 41.48493703 54.68089246 33.        ]
 [19.02629759 34.24908758 14.49137675 45.08880127 69.18092223]]
Время выполнения 100000 повторений: 0.8523911000229418


In [53]:
from scipy.spatial.distance import cdist
#SciPy

x_t75_sp = [[0, -1, 3], [9, -8, 0], [1, -1, 34], [45, -2, 6], [9, -4, -7]]
y_t75_sp = [[16, 9, 5], [26, -26, 13], [4, 4, 4], [0, 0, 37], [77, 5, 2]]

print("SciPy")
print(f"Объеты выборки X: {x_t75_sp}")
print(f"Объекты выборки Y: {y_t75_sp}")

def euclidean_distance_sp(x_objects, y_objects):
    return cdist(x_objects, y_objects, 'euclidean')

print(f"Матрица евклидовых расстояний:\n {euclidean_distance_sp(x_t75_sp, y_t75_sp)}")
print(f"Время выполнения 100000 повторений: {timeit.timeit(lambda: euclidean_distance_sp(x_t75_sp, y_t75_sp), number= 100000)}")

SciPy
Объеты выборки X: [[0, -1, 3], [9, -8, 0], [1, -1, 34], [45, -2, 6], [9, -4, -7]]
Объекты выборки Y: [[16, 9, 5], [26, -26, 13], [4, 4, 4], [0, 0, 37], [77, 5, 2]]
Матрица евклидовых расстояний:
 [[18.97366596 37.42993454  6.4807407  34.0147027  77.23988607]
 [19.05255888 27.96426291 13.60147051 38.91015292 69.26037828]
 [34.14674216 41.12177039 30.56141358  3.31662479 82.68010643]
 [31.0322413  31.40063694 41.48493703 54.68089246 33.        ]
 [19.02629759 34.24908758 14.49137675 45.08880127 69.18092223]]
Время выполнения 100000 повторений: 0.8437571999966167


* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [29]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [30]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [42]:
print(f"Среднее количество калорий в порции конкурентов: {np.mean(calorie_stats)}.")
print(f"\nВ среднем количество калорий в порции конкурентов больше на {np.abs(60 - np.mean(calorie_stats))}, чем в порции CrunchieMunchies")

Среднее количество калорий в порции конкурентов: 106.88311688311688.

В среднем количество калорий в порции конкурентов больше на 46.883116883116884, чем в порции CrunchieMunchies


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [41]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [44]:
print(f"Среднее количество калорий в порции конкурентов: {np.mean(calorie_stats)}.")
print(f"Медиана по количествам калорий в порциях конкурентов: {np.median(calorie_stats)}.")

Среднее количество калорий в порции конкурентов: 106.88311688311688.
Медиана по количествам калорий в порциях конкурентов: 110.0.


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [64]:
nth_percentile = np.min(np.where(np.percentile(calorie_stats, np.arange(0, 101)) > 60))
print(nth_percentile)

4


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [65]:
more_calories = np.mean(calorie_stats > 60) * 100
print(f"У {more_calories:.3f}% конкурентов порция хлопьев калорийнее 60 калорий (калорийнее CrunchieMunchies).")

У 96.104% конкурентов порция хлопьев калорийнее 60 калорий (калорийнее CrunchieMunchies).


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [68]:
calorie_std = np.std(calorie_stats)
print(calorie_std)

19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Полученные данные говорят о следующем
1. Большинство конкурентов (96% +) имеют больше калорий, чем CrunchieMunchies
2. Большое стандратное отклонение показывает то, что данные в выборке достаточно сильно разнятся, но, учитывая вышеописанный пункт, этот факт наоборот приобретает положительный окрас, т.к. отклонение в сторону, которая больше среднего по выборке, является недостатком конкурентов.
3. Разница калорийности CrunchieMunchies и медианы по выборке составляет 83% от калорийности CrunchieMunchies, что тоже можно упомянуть в виде преимущества.